File: 프로그래머스/unrated/176962. 과제 진행하기/과제 진행하기.py
def solution(plans):
    answer = []
    stack = []
    for i in range(len(plans)):
        plans[i][1] = int(plans[i][1][0:2]) * 60 + int(plans[i][1][3:])
        plans[i][2] = plans[i][1] + int(plans[i][2])
    plans.sort(key=lambda x: x[1])
    for i in range(len(plans)):
        name, start_min, end_min = plans[i]
        if i == 0:
            homework = (name, end_min)
            continue
        if homework[1] <= start_min:
            answer.append(homework[0])
            if homework[1] < start_min:
                left_min = start_min - homework[1]
                while stack:
                    last_name, last_time = stack.pop()
                    left_min -= last_time
                    if left_min < 0:
                        stack.append((last_name, abs(left_min)))
                        break
                    elif left_min >= 0:
                        answer.append(last_name)
            homework = (name, end_min)
        else:
            stack.append((homework[0], homework[1] - start_min))
            homework = (name, end_min)
        if i == len(plans) - 1:
            answer.append(homework[0])
    while stack:
        name, time = stack.pop()
        answer.append(name)
    return answer

File: 프로그래머스/2/176962. 과제 진행하기/과제 진행하기.js
function solution(plans) {
    plans.forEach((plan) => {
        const [startHour, startMin] = plan[1].split(':');
        plan[1] = Number(startHour) * 60 + Number(startMin);
        plan[2] = plan[1] + Number(plan[2]);
    });
    const sortedPlans = plans.sort((a, b) => b[1] - a[1]);
    const queue = [];
    const answer = [];
    let time = 0;
    while (sortedPlans.length !== 0) {
        const [title, start, end] = sortedPlans.pop();
        time = start;
        if (sortedPlans.length > 0) {
            const [nextTitle, nextStart, nextEnd] = sortedPlans[sortedPlans.length - 1];
            if (end > nextStart) {
                queue.push([title, end - nextStart]);
            } else if (end === nextStart) {
                answer.push(title);
            } else {
                answer.push(title);
                let leftTime = nextStart - end;
                while (leftTime > 0 && queue.length > 0) {
                    const [beforeTitle, beforeLeftTime] = queue.pop();
                    if (beforeLeftTime > leftTime) {
                        queue.push([beforeTitle, beforeLeftTime - leftTime]);
                        leftTime = 0;
                    } else {
                        answer.push(beforeTitle);
                        leftTime -= beforeLeftTime;
                    };
                };
            };
        } else {
            answer.push(title, ...queue.reverse().flatMap((homework) => homework[0]));
        };
    };
    
    return answer;
}